File: medium/199. Binary Tree Right Side View/199. Binary Tree Right Side View.go
package main

/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	res := []int{}
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		levelSize := len(queue)
		rightVal := 0

		for i := 0; i < levelSize; i++ {
			node := queue[0]
			queue = queue[1:]

			if i == levelSize-1 {
				rightVal = node.Val
			}

			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		res = append(res, rightVal)
	}

	return res
}

File: medium/199. Binary Tree Right Side View/199. Binary Tree Right Side View.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function rightSideView(root: TreeNode | null): number[] {
    if (root == null) {
        return []
    }

    const res = []
    const queue = [root]

    while (queue.length) {
        let rightVal = 0
        const levelSize = queue.length

        for (let i = 0; i <= levelSize; i++) {
            const node = queue.shift()!

            if (i == levelSize - 1) {
                rightVal = node.val
            }

            if (node.left) {
                queue.push(node.left)
            }
            if (node.right) {
                queue.push(node.right)
            }
        }

        res.push(rightVal)
    }

    return res
};

export { }
